Index: test/utils.py
===================================================================
--- test/utils.py	(revision 7671)
+++ test/utils.py	(working copy)
@@ -74,7 +74,10 @@
 def _set_autocommit(connection):
     "Make sure a connection is in autocommit mode."
     if hasattr(connection.connection, "autocommit"):
-        connection.connection.autocommit(True)
+        if callable(connection.connection.autocommit):
+            connection.connection.autocommit(True)
+        else:
+            connection.connection.autocommit = True
     elif hasattr(connection.connection, "set_isolation_level"):
         connection.connection.set_isolation_level(0)
 
Index: contrib/sessions/backends/db.py
===================================================================
--- contrib/sessions/backends/db.py	(revision 7671)
+++ contrib/sessions/backends/db.py	(working copy)
@@ -13,9 +13,12 @@
 
     def load(self):
         try:
+            date_now = datetime.datetime.now()
+            if hasattr(date_now, 'microsecond'):
+                date_now = date_now.replace(microsecond=0)
             s = Session.objects.get(
                 session_key = self.session_key,
-                expire_date__gt=datetime.datetime.now()
+                expire_date__gt=date_now
             )
             return self.decode(s.session_data)
         except (Session.DoesNotExist, SuspiciousOperation):
